src/GafferUI/FPSGadget.cpp

```diff
@@ -0,0 +1,115 @@
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferUI/FPSGadget.h"
+
+#include "GafferUI/Style.h"
+#include "GafferUI/ViewportGadget.h"
+
+using namespace GafferUI;
+
+GAFFER_GRAPHCOMPONENT_DEFINE_TYPE( FPSGadget );
+
+FPSGadget::FPSGadget()
+	:	Gadget( defaultName<FPSGadget>() ), m_totalTime( 0 )
+{
+}
+
+FPSGadget::~FPSGadget()
+{
+}
+
+void FPSGadget::renderLayer( Gadget::Layer layer, const Style *style, Gadget::RenderReason reason ) const
+{
+	if( layer != Layer::Main )
+	{
+		return;
+	}
+
+	if( isSelectionRender( reason ) )
+	{
+		return;
+	}
+
+	static std::chrono::time_point<std::chrono::high_resolution_clock> lastTime;
+	static bool initialized = false;
+
+	auto curTime = std::chrono::high_resolution_clock::now();
+	int frameTime = std::chrono::duration_cast< std::chrono::microseconds >( curTime - lastTime ).count();
+	lastTime = curTime;
+
+	if( !initialized )
+	{
+		initialized = true;
+		return;
+	}
+
+	// Simple running average
+	m_totalTime += frameTime;
+	m_timeBuffer.push_back( frameTime );
+
+	// 20 frames is enough to smooth out bumps, while not lagging behind too much
+	while( m_timeBuffer.size() > 20 )
+	{
+		m_totalTime -= m_timeBuffer.front();
+		m_timeBuffer.pop_front();
+	}
+
+	float averageFrameTime = m_totalTime / ( m_timeBuffer.size() * 1000000.0f );
```

Comment:
I've tended to use an exponential moving average for this sort of thing in the past, because it doesn't need a buffer : https://stackoverflow.com/a/4687507. Is there any downside to that?

Reply:
My thought was that it's important to have a definite period of time over which the old values are definitely completely gone, particularly because unlike in a lot of things with FPS counters, it's possible to have extremely slow updates in Gaffer ( because nothing may trigger an update for an extended period of time ).

The buffer size of 20 is actually a pretty awkward compromise between stability and updating quickly enough when the FPS is low.

I realized last night that I could do something much more reliable: average over the past 1 second.  If the update has taken longer than a second, then the FPS only considers the last frame.  I think my idea for implementing that may look a little cleaner anyway.

Reply:
> it's possible to have extremely slow updates in Gaffer ( because nothing may trigger an update for an extended period of time )

I did have one thought on that but forgot to mention it on the PR. Perhaps we could/should uninitialise `lastTime` on `visibilityChangedSignal()` so that an extended period of not being drawn due to invisibility doesn't throw off the initial readouts? Your point would still stand about periods where the gadget is visible but not needing a redraw for other reasons though.

> I realized last night that I could do something much more reliable: average over the past 1 second. If the update has taken longer than a second, then the FPS only considers the last frame. I think my idea for implementing that may look a little cleaner anyway.

Cool, sounds worth a try. Let's try to get this wrapped up quickly though - I think any of the approaches is probably sufficient for our purposes, and there's plenty else we need to move on to.